File: Shared_Files/Jaxyn_Code/gaussian_fitter.m
clear; close all; clc
global frame_to_fit xp yp deltax deltay ifit iter;

% frame = open('J.mat'); 
frame = open('../dot_grid.mat');
square_length = 10;
half = round(square_length/2);
subframe = double(frame.fits(930-half:930+half,810-half:810+half));
% subframe = frame.J(1025-round(square_length/2):1025+round(square_length/2),2185-round(square_length/2):2185+round(square_length/2));
frame_to_fit = subframe; % col 27 of subframe for PSF
initialPars = [300, 15000,0,square_length/2,square_length/2,1,1];
[xp,yp] = meshgrid(1:square_length+1);
options = LMFnlsq;
options = LMFnlsq(options,'FunTol',1e-7,'XTol',1e-7);
ifit = 1; iter = 0;
[par_frameFit, ssq,cnt,nfJ, ~] = LMFnlsq('rotated_gaussian',initialPars,options);

[res, fitted_gaussian] = rotated_gaussian(par_frameFit);
disp(par_frameFit);

figure;
imagesc(subframe);colorbar;
figure;
% look at residual
imagesc(subframe-fitted_gaussian);colorbar;

sigmax = abs(par_frameFit(6));
sigmay = abs(par_frameFit(7));
a=sigmax; % horizontal radius
b=sigmay; % vertical radius
x0=par_frameFit(4); % x0,y0 ellipse centre coordinates
y0=par_frameFit(5);
% rotate the ellipse according to the fitted orientation angle
theta = par_frameFit(3);
X = cos(theta)*(xp-x0) - sin(theta)*(yp-y0);
Y = sin(theta)*(xp-x0) + cos(theta)*(yp-y0);
% get the pixels that fall closest to the ellipses 
%ellipse_pixels1 = (X.^2/a^2 + Y.^2/b^2 > 0.55 & X.^2/a^2 + Y.^2/b^2 < 1.45); 
%ellipse_pixels2 = (X.^2/(2*a)^2 + Y.^2/(2*b)^2 > 0.75 & X.^2/(2*a)^2 + Y.^2/(2*b)^2 < 1.35); 
%ellipse_pixels3 = (X.^2/(3*a)^2 + Y.^2/(3*b)^2 > 0.85 & X.^2/(3*a)^2 + Y.^2/(3*b)^2 < 1.25); 
% actually, just shade in the regions
ellipse_pixels1 = (X.^2/a^2 + Y.^2/b^2 <= 1); 
ellipse_pixels2 = (X.^2/(2*a)^2 + Y.^2/(2*b)^2 <= 1); 
ellipse_pixels3 = (X.^2/(3*a)^2 + Y.^2/(3*b)^2 <=1); 

figure;
imagesc(fitted_gaussian);colorbar;
figure;
subframe1 = subframe;
subframe1(ellipse_pixels1) = inf; % color the ellipses to show up on figure
imagesc(subframe1);colorbar;
figure;
subframe2 = subframe;
subframe2(ellipse_pixels2) = inf;
imagesc(subframe2);colorbar;
figure;
subframe3 = subframe;
subframe3(ellipse_pixels3) = inf;
imagesc(subframe3);colorbar;


% normalized power spectral density of PSF 
PSF_col = subframe(:,half+1);   % column 27 of subframe for this particular PSF
s = size(PSF_col);    
% Length of signal (minus 1 if necessary to make the number even)
L = s(1) -1; 
Fs = 1;     % Sampling frequency: 1 per pixel is all we have                    
f = Fs/L*(0:(L/2));  % frequency domain, just the positive range
% FFT:
Y = abs(fft(PSF_col));
% divide by length of signal which scaled up the signal via fft, and
% multiply by 2 since the amplitude was split across the positives and
% negatives.  Then square it to get the power spectral amplitude
Y = (2*Y/L).^2;
% shift FFT to display positive and negative frequencies in order, and take
% absolute value for real part
Ypos = Y(1:floor(L/2)+1);
figure; semilogy(f, Ypos/sum(abs(Ypos)));

return

%%%%%%%%%%%%%%% Functions Used %%%%%%%%%%%%%%%

function img = clnShow(array,numSigs,optHead,optAxis)
% Input an array, and see it cleanly displayed on a new figure, with no 
% influence on color range by outliers. If you want to add a title, give
% that as the second input. If you want to do things with the axis, such as
% "axis equal" or such, give whatever argument you want to put in axis() as
% the third argument. (If you want to give an axis argument but don't want
% a title, simply give an empty char array, '', as the second argument.)
% ​
if nargin <2
    numSigs = 3;
end
if nargin < 3
    optHead = '';
end
if nargin < 4
    optAxis = "on";
end
statArray = array(:);
preMean = mean(statArray);
stdev = std(statArray);
lowNufs = statArray <= preMean + numSigs*stdev;
lowNuf = statArray(lowNufs);
highNufs = lowNuf >= lowNuf - numSigs*stdev;
clean = lowNuf(highNufs);
avg = mean(clean);
sdev = std(clean);

val = avg - numSigs*sdev;
pek = avg + numSigs*sdev;

figure
img = imagesc(array,[val,pek]);
colorbar; img.Tag = optHead; title(optHead);
axis(optAxis)
end

function y = nPeaks(M_in, n, m)
% find the n highest peaks, assuming that the no other peak exists within 
% m pixels of each found peak 
% B. Nemati, 15-May-2012

[nr,nc] = size(M_in);
M = M_in;
minM = min(min(M));
for ip = 1:n
    maxM = max(max(M));
    [ir, ic] = find(M == maxM, 1, 'first');
    M(max(ir-m,1):min(ir+m,nr),max(ic-m,1):min(ic+m,nc)) = minM;
    peak(ip).value = maxM;
    peak(ip).row = ir;
    peak(ip).col = ic;
end
y = peak;
end

File: Shared_Files/gaussian_fitter.py
"""
Python conversion of gaussian_fitter.m

https://claude.ai/chat/821bf681-6e20-4bc5-a676-f6cc65fd18ae

Notes on the conversion
------------------------
- MATLAB's LMFnlsq (a third-party Levenberg-Marquardt solver) is replaced
  with scipy.optimize.least_squares(..., method='lm'), which implements
  the same underlying algorithm (MINPACK's lmdif), with FunTol/XTol
  mapped to ftol/xtol.
- MATLAB uses 1-based, inclusive-end array indexing; the subframe crop
  below has been converted to 0-based, exclusive-end Python slicing that
  selects the identical set of pixels.
- MATLAB globals (frame_to_fit, xp, yp, ifit, iter) are replaced with
  ordinary local variables that get passed explicitly into
  rotated_gaussian / rotated_gaussian_residuals.
- imagesc(...) -> plt.imshow(..., origin='upper') with a colorbar.
- The helper functions clnShow and nPeaks (defined but unused after the
  `return` in the original .m file) are included at the bottom in
  Python form for completeness/parity, faithfully preserving the
  original logic (including an apparent no-op comparison in clnShow's
  `highNufs` line, which is always true as written in the source).
"""

import numpy as np
from scipy.optimize import least_squares
import matplotlib.pyplot as plt
import os

import helper_functions.file_io as file_io
from helper_functions.rotated_gaussian import *



def show_results(subframe: np.ndarray, fitted_gaussian, par_frame_fit: tuple, xp, yp):

    plt.figure()
    plt.imshow(subframe)
    plt.colorbar()

    plt.figure()
    # look at residual
    plt.imshow(subframe - fitted_gaussian)
    plt.colorbar()

    square_length = subframe.shape[0]
    x0 = par_frame_fit[3]  # ellipse centre coordinates
    y0 = par_frame_fit[4]
    sigmax = abs(par_frame_fit[5]) # horizontal radius
    sigmay = abs(par_frame_fit[6]) # vertical radius

    # rotate the ellipse according to the fitted orientation angle
    theta = par_frame_fit[2]
    X = np.cos(theta) * (xp - x0) - np.sin(theta) * (yp - y0)
    Y = np.sin(theta) * (xp - x0) + np.cos(theta) * (yp - y0)

    # get the pixels that fall closest to the ellipses (shade in the regions)
    ellipse_pixels1 = (X ** 2 / sigmax ** 2 + Y ** 2 / sigmay ** 2 <= 1)
    ellipse_pixels2 = (X ** 2 / (2 * sigmax) ** 2 + Y ** 2 / (2 * sigmay) ** 2 <= 1)
    ellipse_pixels3 = (X ** 2 / (3 * sigmax) ** 2 + Y ** 2 / (3 * sigmay) ** 2 <= 1)

    plt.figure()
    plt.imshow(fitted_gaussian)
    plt.colorbar()

    plt.figure()
    subframe1 = subframe.copy()
    subframe1[ellipse_pixels1] = np.inf  # color the ellipses to show up on figure
    plt.imshow(subframe1)
    plt.colorbar()

    plt.figure()
    subframe2 = subframe.copy()
    subframe2[ellipse_pixels2] = np.inf
    plt.imshow(subframe2)
    plt.colorbar()

    plt.figure()
    subframe3 = subframe.copy()
    subframe3[ellipse_pixels3] = np.inf
    plt.imshow(subframe3)
    plt.colorbar()

    # normalized power spectral density of PSF
    psf_idx = np.argmax(subframe)
    psf_y, psf_x = np.unravel_index(psf_idx, subframe.shape)
    PSF_col = subframe[:, psf_x]  # column 14 (1-based) of subframe for this particular PSF
    L = PSF_col.shape[0]   # length of signal (minus 1 to make it even, if needed)
    Fs = 1  # sampling frequency: 1 per pixel is all we have
    f = Fs / L * np.arange(0, L / 2 + 1)  # frequency domain, positive range only

    Y_fft = np.abs(np.fft.fft(PSF_col))
    # divide by length of signal (fft scaling), multiply by 2 (amplitude split
    # across +/- frequencies), then square for power spectral amplitude
    Y_fft = (2 * Y_fft / L) ** 2
    Ypos = Y_fft[:int(np.floor(L / 2)) + 1]

    plt.figure()
    plt.semilogy(f, Ypos / np.sum(np.abs(Ypos)))

    plt.show()

def fit_gaussian(image: np.ndarray, square_length: int, psf_x: int, psf_y: int, verbose=True) -> tuple:

    # frame = loadmat('J.mat')
    # frame = loadmat('Shared_Files/dot_grid.mat')
    # image = frame['fits']
    half = square_length // 2

    # row_slice = matlab_range_to_slice(psf_x - half, psf_x + half)
    # col_slice = matlab_range_to_slice(psf_y - half, psf_y + half)
    if square_length / 2 != half:
        row_slice = slice(psf_y - half - 1, psf_y + half)
        col_slice = slice(psf_x - half - 1, psf_x + half)
    else:        
        row_slice = slice(psf_y - half, psf_y + half)
        col_slice = slice(psf_x - half, psf_x + half)

    subframe = image[row_slice, col_slice].astype(float)
    # subframe = frame['J'][matlab_range_to_slice(1025-half,1025+half),
    #                       matlab_range_to_slice(2185-half,2185+half)]

    frame_to_fit = subframe  # col 27 (index 26) of subframe holds the PSF

    initial_pars = np.array([300, 15000, 0, half, half, 1, 1], dtype=float)

    xp, yp = np.meshgrid(np.arange(0, square_length), np.arange(0, square_length))

    ifit = 1

    result = least_squares(
        rotated_gaussian_residuals,
        initial_pars,
        args=(xp, yp, frame_to_fit),
        kwargs={'ifit': ifit, 'verbose': verbose},
        method='lm',
        ftol=1e-7,
        xtol=1e-7,
    )
    par_frame_fit = result.x
    par_frame_fit[2] = np.mod(par_frame_fit[3], 2*np.pi)

    _, fitted_gaussian = rotated_gaussian(par_frame_fit, xp, yp, frame_to_fit, ifit=ifit, verbose=False)
    if verbose:
        print(par_frame_fit)

    return(subframe, fitted_gaussian, par_frame_fit, xp, yp)

def main():

    psf_x = 720
    psf_y = 1119

    file_names, path = file_io.get_directory_input()    

    file_path = os.path.join(path, file_names[0])

    image = file_io.read_fits_file(file_path)

    square_length = 10

    results = fit_gaussian(image, square_length, psf_x, psf_y)

    show_results(*results)

if __name__ == '__main__':
    main()